Empty body for zero length payloads

jamesgorman2 opened the pull request:
Update `RawRequest` and `HttpServerResponseImpl` so they return `content-length: 0` and an empty body when no writes are made to `HttpClientRequest` and `HttpServerResponse` respectively.

This also adds some regression level tests against the bytes sent and received from the server.

 (#470, #520)

jamesgorman2 committed:
Empty body for zero length payloads (#470, #520)

Update `RawRequest` and `HttpServerResponseImpl` so they return `content-length: 0` and an empty body
when no writes are made to `HttpClientRequest` and `HttpServerResponse` respectively.

This also adds some regression level tests against the bytes sent and received from the server.

jamesgorman2 commented on rxnetty-http/src/main/java/io/reactivex/netty/protocol/http/client/internal/RawRequest.java:
```diff
@@ -166,6 +166,12 @@ private RawRequest(HttpRequest headers, Observable content, Func1<?, Boolean> fl
+            headers.headers().set(HttpHeaderNames.CONTENT_LENGTH, 0);
+        }
+
         Observable toReturn = Observable.just(headers);
```
This is cognate to `ContentWriterImpl` ll61

jamesgorman2 commented on rxnetty-http/src/main/java/io/reactivex/netty/protocol/http/server/HttpServerResponseImpl.java:
```diff
@@ -52,7 +52,7 @@ private HttpServerResponseImpl(final State<C> state) {
             public void call(Subscriber<? super Void> subscriber) {
-                state.sendHeaders().write(Observable.<C>empty()).unsafeSubscribe(subscriber);
+                state.sendHeaders().unsafeSubscribe(subscriber);
             }
```
Turns out this is `write` unnecessary - this was causing differing behaviour between 

```
public Observable<Void> handle(HttpServerRequest<ByteBuf> request, HttpServerResponse<ByteBuf> response) {
    return response;
}
```

and

```
public Observable<Void> handle(HttpServerRequest<ByteBuf> request, HttpServerResponse<ByteBuf> response) {
    return response
        .sendHeaders();
}
```

the former previously returned with an empty chunk, the latter `content-length: 0`

jamesgorman2 commented on rxnetty-http/src/test/java/io/reactivex/netty/protocol/http/client/HttpClientTest.java:
```diff
@@ -272,6 +274,101 @@ public void call() {
+                @Override
+                public Observable<HttpClientResponse<ByteBuf>> call(HttpClient<ByteBuf, ByteBuf> client) {
+                    return client.createGet("/")
+                        .setHeader(HttpHeaderNames.CONTENT_LENGTH, 5);
```
This is to match behaviour on the server side

jamesgorman2 commented on rxnetty-http/src/test/java/io/reactivex/netty/protocol/http/server/HttpServerTest.java:
```diff
@@ -0,0 +1,131 @@
+    public final HttpServerRule serverRule = new HttpServerRule();
+
+    @Test(timeout = 60000)
+    public void testResponseWithNoContentLengthHeaderOrContentReturnsEmptyBody() throws Exception {
```
These two are the regressions for calling/calling not calling `#sendHeaders()`

NiteshKant commented on rxnetty-http/src/test/java/io/reactivex/netty/protocol/http/client/HttpClientTest.java:
```diff
@@ -272,6 +274,101 @@ public void call() {
+                    return client.createGet("/");
+                }
+            },
+            "GET / HTTP/1.1\r\n" +
```
Not sure if checking raw HTTP message is required, can you pass a custom handler that asserts request headers and content inside the server and store the error/success to be queried later?

jamesgorman2 replied:
I didn't see the code in `HttpClientRule` for this the first time round (I started from the server rule). I've push a commit to address this that can be squashed in if it's good.

NiteshKant commented:
@jamesgorman2 Thanks for this, good work.

When you address my review comment, I will be happy to pull this in.

jamesgorman2 committed:
Clean up tests

Make in style of other tests byt hiding details.

Move client test code over to client rule

TO BE SQUASHED

jamesgorman2 commented on rxnetty-http/src/test/java/io/reactivex/netty/protocol/http/server/HttpServerRule.java:
```diff
@@ -114,4 +183,55 @@ public void setServer(HttpServer<ByteBuf, ByteBuf> server) {
+    }
+
+
+    private static class RawMessageHandler extends ChannelDuplexHandler {
```
I've left this in to capture the response since there's no `#channelProvider()` method on the server side.

NiteshKant commented:
Thanks @jamesgorman2 !
